src/diffusers/models/adapter.py

```diff
@@ -0,0 +1,307 @@
+                            block_out_channels[i - 1],
+                            block_mid_channels[i],
+                            block_mid_channels[i],
+                            down=True,
+                            ksize=kerenl_size,
+                            proj_ksize=proj_kerenl_size,
+                            sk=res_block_skip,
+                            use_conv=use_conv,
+                        )
+                    )
+                elif j == num_res_blocks - 1:
+                    self.body.append(
+                        ResnetBlock(
+                            block_mid_channels[i],
+                            block_mid_channels[i],
+                            block_out_channels[i],
+                            down=False,
+                            ksize=kerenl_size,
+                            proj_ksize=proj_kerenl_size,
+                            sk=res_block_skip,
+                            use_conv=use_conv,
+                        )
+                    )
+                else:
+                    self.body.append(
+                        ResnetBlock(
+                            block_mid_channels[i],
+                            block_mid_channels[i],
+                            block_mid_channels[i],
+                            down=False,
+                            ksize=kerenl_size,
+                            proj_ksize=proj_kerenl_size,
+                            sk=res_block_skip,
+                            use_conv=use_conv,
+                        )
+                    )
+        self.body = nn.ModuleList(self.body)
+
+        if block_mid_channels[0] == block_out_channels[0]:
+            # follow standar adapter schema, using fix kernel size 3 for stem conv layer
+            self.conv_in = nn.Conv2d(channels_in * input_scale_factor**2, block_mid_channels[0], 3, 1, 1)
+        else:
+            # if block_mid_channels[i] < block_out_channels[i](bottleneck downsample block), using light weight adapter schema instead
+            self.conv_in = nn.Conv2d(
+                channels_in * input_scale_factor**2,
+                block_mid_channels[0],
+                proj_kerenl_size,
+                1,
+                proj_kerenl_size // 2,
+            )
+
+    def forward(self, x: torch.Tensor) -> List[torch.Tensor]:
+        r"""
+        Args:
+            x (`torch.Tensor`):
+                (batch, channel, height, width) input images for adapter model, `channel` should equal to
+                `channels_in`.
+        """
+        # unshuffle
+        x = self.unshuffle(x)
+        # extract features
+        features = []
+        x = self.conv_in(x)
+        for i in range(self.num_downsample_blocks):
+            for j in range(self.num_res_blocks):
+                idx = i * self.num_res_blocks + j
+                x = self.body[idx](x)
+            features.append(x)
+
+        return features
+
+
+class MultiAdapter(ModelMixin, ConfigMixin):
```

Comment:
Nice!

Can we make MultiAdapter follow MultiControlnet ? https://github.com/huggingface/diffusers/blob/a16957159e24e53df0a40bfadbcd0a61b4b3ea8c/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_controlnet.py#L92

It shouldn't need to construct its own classes internally. Its constructor should always only take a list of adapter models

the adapter weights can always be passed to the forward function instead of in the constructor

Because it always only takes pre-constructed Adapter models, it doesn't need to use the config mixin

Reply:
The current design is for supporting something like `MultiAdapter.from_pretrained"RzZ/sd-v1-4-adapter-keypose-depth")`, which will also import the `adpater_weights` form checkpoint. I think it will be useful when users want to share a more complicated adapters setup.

Reply:
If we need to support more complicated adapters in the future we can add alternative model definitions then, let's follow the control net example for now :)